ocaml/forkexecd/lib/fe_stubs.c

```diff
@@ -0,0 +1,899 @@
+                if (op->fd_from == op->fd_to) {
+                    err_func = "fcntl";
+                    int flags = fcntl(op->fd_from, F_GETFD);
+                    if (flags == -1) {
+                        err = errno;
+                    } else if ((flags & FD_CLOEXEC) != 0) {
+                        if (fcntl(op->fd_from, F_SETFD, flags & ~FD_CLOEXEC) == -1)
+                            err = errno;
+                    }
+                } else {
+                    err_func = "dup2";
+                    if (dup2(op->fd_from, op->fd_to) < 0)
+                        err = errno;
+                }
+                break;
+            case FD_OP_MOVE:
+                err_func = "dup2";
+                if (dup2(op->fd_from, op->fd_to) < 0)
+                    err = errno;
+                close(op->fd_from);
+                break;
+            case FD_OP_CLOSE:
+                close(op->fd_from);
+                break;
+            case FD_OP_DEVNULL:
+                // first close old, then create new one
+                close(op->fd_to);
+                // TODO ideally we want read only for input for Ocaml did the same...
+                err_func = "open";
+                if (open("/dev/null", O_WRONLY) != op->fd_to)
+                    err = errno ? errno : EBADF;
+                break;
+            case FD_OP_CLOSE_FROM:
+                close_fds_from(op->fd_from);
+                break;
+            default:
+                err_func = "safe_exec";
+                err = EINVAL;
+            }
+        }
+        if (err != 0) {
+            info->err = err;
+            info->err_func = err_func;
+            log_fail("redirect error %d in %s", err, err_func);
+            _exit(125);
+        }
+
+        // Limit number of files limits to standard limit to avoid
+        // creating bugs with old programs.
+        if (nofile_limit.rlim_cur > 1024) {
+            nofile_limit.rlim_cur = 1024;
+            setrlimit(RLIMIT_NOFILE, &nofile_limit);
+        }
+
+        // Reset signal mask, inherited by the process we are going to execute
+        sigemptyset(&sigset);
+        pthread_sigmask(SIG_SETMASK, &sigset, NULL);
+
+        log("execve...");
+        mapped_logs_success(logs);
+        execve(info->args[0], info->args, info->environment);
+        log_fail("execve failed %d", errno);
+        // Here we could set err and err_func but we kept compatibility
+        // with forkexecd daemon.
+        _exit(errno == ENOENT ? 127 : 126);
+    }
+
+    log("done");
+
+    // Restore thread state
+    pthread_sigmask(SIG_SETMASK, &old_sigset, NULL);
+    pthread_setcancelstate(cancellation_state, NULL);
+    mapped_logs_close(logs);
+
+    }
+    caml_leave_blocking_section();
+
+    // Handle errors from child.
+    int err = info->err;
+    const char *err_func = info->err_func;
+    caml_stat_free(info);
+    if (err != 0 && err_func != NULL) {
+        while (waitpid(pid, NULL, 0) < 0 && errno == EINTR)
```

Comment:
This calls a blocking kernel function, but we're already holding the OCaml runtime lock, is the call in the wrong place?

Reply:
This is just reaping a zombie process, not blocking